crates/typst/src/export/render.rs

```diff
@@ -626,12 +685,97 @@ impl From<Transform> for sk::Transform {
-        sk_paint.set_color(color.into());
-        sk_paint.anti_alias = true;
+impl From<sk::Transform> for Transform {
+    fn from(value: sk::Transform) -> Self {
+        let sk::Transform { sx, ky, kx, sy, tx, ty } = value;
+        Self {
+            sx: Ratio::new(sx as _),
+            ky: Ratio::new(ky as _),
+            kx: Ratio::new(kx as _),
+            sy: Ratio::new(sy as _),
+            tx: Abs::raw(tx as _),
+            ty: Abs::raw(ty as _),
+        }
+    }
+}
+
+/// Transforms a [`Paint`] into a [`sk::Paint`].
+/// Applying the necessary transform, if the paint is a gradient.
+trait IntoSkPaint {
+    fn as_sk_paint<'a>(
+        &self,
+        state: &State,
+        item_size: Size,
+        fill_transform: Option<sk::Transform>,
+        pixmap: &'a mut Option<Arc<sk::Pixmap>>,
+    ) -> sk::Paint<'a>;
+}
+
+impl IntoSkPaint for Paint {
+    fn as_sk_paint<'a>(
+        &self,
+        state: &State,
+        item_size: Size,
+        fill_transform: Option<sk::Transform>,
+        pixmap: &'a mut Option<Arc<sk::Pixmap>>,
+    ) -> sk::Paint<'a> {
+        /// Actual sampling of the gradient, cached for performance.
+        #[comemo::memoize]
+        fn cached(gradient: &Gradient, width: u32, height: u32) -> Arc<sk::Pixmap> {
+            let mut pixmap = sk::Pixmap::new(width.max(1), height.max(1)).unwrap();
+            for x in 0..width {
+                for y in 0..height {
+                    let color: sk::Color = gradient
+                        .sample_at((x as f32, y as f32), (width as f32, height as f32))
+                        .into();
+
+                    pixmap.pixels_mut()[(y * width + x) as usize] =
+                        color.premultiply().to_color_u8();
+                }
+            }
+
+            Arc::new(pixmap)
+        }
+
+        let mut sk_paint: sk::Paint<'_> = sk::Paint::default();
+        match self {
+            Paint::Solid(color) => {
+                sk_paint.set_color((*color).into());
+                sk_paint.anti_alias = true;
+            }
+            Paint::Gradient(gradient) => {
+                let container_size = match gradient.unwrap_relative(false) {
+                    Relative::This => item_size,
+                    Relative::Parent => state.size,
+                };
+
+                let fill_transform = fill_transform.unwrap_or_else(|| {
+                    match gradient.unwrap_relative(false) {
+                        Relative::This => sk::Transform::identity(),
+                        Relative::Parent => state
+                            .container_transform
+                            .post_concat(state.transform.invert().unwrap()),
+                    }
+                });
+                let width =
+                    (container_size.x.to_f32() * state.pixel_per_pt).ceil() as u32;
+                let height =
+                    (container_size.y.to_f32() * state.pixel_per_pt).ceil() as u32;
```

Comment:
Is there any risk of overflow here? (Should perhaps be documented, or handled)

Reply:
If there is then it will happen (much) sooner in the rendering code, so it's a non issue for this specific case. That being said, it should be handled earlier in the chain since currently I'm pretty sure it will panic (should be another PR)

Reply:
float -> int casts saturate as far as I know